test/nestedJoin.integration.test.ts

```diff
@@ -0,0 +1,432 @@
+    parseUnits,
+    publicActions,
+    PublicActions,
+    TestActions,
+    WalletActions,
+    walletActions,
+} from 'viem';
+
+import {
+    Slippage,
+    NestedJoin,
+    Token,
+    replaceWrapped,
+    NestedPoolState,
+} from '../src/entities';
+import { Address } from '../src/types';
+
+import {
+    BALANCER_RELAYER,
+    BALANCER_VAULT,
+    CHAINS,
+    ChainId,
+} from '../src/utils';
+
+import {
+    approveToken,
+    findTokenBalanceSlot,
+    sendTransactionGetBalances,
+    setTokenBalance,
+} from './lib/utils/helper';
+import { authorizerAbi, vaultAbi } from '../src/abi';
+import { Relayer } from '../src/entities/relayer';
+import { NestedJoinInput } from '../src/entities/nestedJoin/types';
+
+/**
+ * Deploy the new relayer contract with the new helper address:
+ *
+ * in the mono repo:
+ * cd pkg/standalone-utils
+ * forge create --rpc-url http://0.0.0.0:8545 --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 contracts/BatchRelayerQueryLibrary.sol:BatchRelayerQueryLibrary --constructor-args "0xBA12222222228d8Ba445958a75a0704d566BF2C8"
+ *
+ * [take the address]
+ *
+ * forge create --rpc-url http://0.0.0.0:8545 --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 contracts/relayer/BalancerRelayer.sol:BalancerRelayer --constructor-args "0xBA12222222228d8Ba445958a75a0704d566BF2C8" "0xf77018c0d817da22cadbdf504c00c0d32ce1e5c2" "[paste the address]" "5"
+ *
+ * update `BALANCER_RELAYER` on constants.ts
+ *
+ */
+
+type TxInput = {
+    amountsIn: {
+        address: Address; // DAI
+        rawAmount: bigint;
+        decimals: number;
+    }[];
+    chainId: ChainId;
+    rpcUrl: string;
+    testAddress: Address;
+    nestedJoin: NestedJoin;
+    nestedPoolFromApi: NestedPoolState;
+    client: Client & PublicActions & TestActions & WalletActions;
+    useNativeAssetAsWrappedAmountIn?: boolean;
+};
+
+describe('nested join test', () => {
+    let api: MockApi;
+    let chainId: ChainId;
+    let rpcUrl: string;
+    let client: Client & PublicActions & TestActions & WalletActions;
+    let poolAddress: Address;
+    let nestedPoolFromApi: NestedPoolState;
+    let nestedJoin: NestedJoin;
+    let testAddress: Address;
+
+    beforeAll(async () => {
+        // setup mock api
+        api = new MockApi();
+
+        // setup chain and test client
+        chainId = ChainId.MAINNET;
+        rpcUrl = 'http://127.0.0.1:8545/';
+        client = createTestClient({
+            mode: 'hardhat',
+            chain: CHAINS[chainId],
+            transport: http(rpcUrl),
+        })
+            .extend(publicActions)
+            .extend(walletActions);
+
+        testAddress = (await client.getAddresses())[0];
+
+        poolAddress = '0x08775ccb6674d6bdceb0797c364c2653ed84f384'; // WETH-3POOL-BPT
+
+        // get pool state from api
+        nestedPoolFromApi = await api.getNestedPool(poolAddress);
+
+        // setup join helper
+        nestedJoin = new NestedJoin();
```

Comment:
I normally find tests are a good place to see how functionality can be used. Would be good to keep all the test setup, etc together (or even better in a separate reusable helper) and the core functionality before/after. Similar for doTransaction. Does that make sense?

Reply:
I agree. I'll refactor it over here.